File: tokit/tokit/language/xsd/xsd_generator.cpp
///<------------------------------------------------------------------------------
//< @file:   prop\parser\gen_xsd.cpp
//< @author: 洪坤安
//< @date:   2014年4月5日 17:39:38
//< @brief:  
//< Copyright (c) 2014 Tokit. All rights reserved.
///<------------------------------------------------------------------------------

#include "xsd_generator.h"

#include <fstream>

#include "cfg.h"
#include "file_util.h"
#include "str_util.h"
#include "echoutil.h"
#include "parser.h"

namespace xsdutil{
    void gen_head(ofstream &o, const cfg_t &cfg)
    {
	    o << "<?xml version=\"1.0\" encoding=\"UTF-8\"?>" << endl;
    }

    void gen_elem(ofstream &o, const cfg_t &cfg)
    {
	    o << "<xs:element name=\"" << cfg.en_name << "s\">" << std::endl;
	    o << "  <xs:complexType>" << endl;
	    o << "    <xs:sequence minOccurs=\"0\" maxOccurs=\"unbounded\">" << endl;
	    o << "      <xs:element name=\"" << cfg.en_name << "\">" << std::endl;
	    o << "        <xs:complexType>" << endl;
	    o << "          <xs:sequence>" << endl;

        /*
        for(size_t n = 0; n < cfg.fields.size(); n++){
            const field_t &field = cfg.fields[n];
		    string& xsd_type = cfgutil::raw_type_2_xsd_type(field.fieldtype);
		    o << "            <xs:element name=\"" << field.fieldname << "\" type=\"" << xsd_type << "\"/>" << endl;
	    }
        */

        for(size_t n = 0; n < cfg.fields.size(); n++){
		    const field_t &field = cfg.fields[n];

		    string& xsd_type = xsd_generator::tokit_type_2_xsd_type(field.fieldtype);
		    o << "            <xs:attribute name=\"" << field.en_name << "\" type=\"" << xsd_type << "\"/>" << endl;
	    }

        o << "          </xs:sequence>" << endl;
	    o << "        </xs:complexType>" << endl;    
	    o << "      </xs:element>" << endl;
	    o << "    </xs:sequence>" << endl;
	    o << "  </xs:complexType>" << endl;    
	    o << "</xs:element>" << endl;
    }

    void gen_body(ofstream &o, const cfg_t &cfg)
    {
	    o << "<xs:schema xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">" << endl;
	    gen_elem(o, cfg);
	    o << "</xs:schema>" << endl;
    }

    void gen_xsd(const cfg_t &cfg, const string &xsd)
    {
	    ofstream o(xsd);

	    gen_head(o, cfg);
	    gen_body(o, cfg);

        ECHO_OK("导出[%s] -> [%s]成功", cfg.cn_name.c_str(), xsd.c_str());
    }
}


bool xsd_generator::generate()
{
    if(false == fileutil::exist(m_xsd_dir)){
        ECHO_ERR("错误: <%s>路径不存在", m_xsd_dir.c_str());
        return false;
    }

    return gen_xsds();
}

bool xsd_generator::gen_xsds()
{
    size_t n_cfg = m_cfgbase.cfgs.size();
    for(size_t n = 0; n < n_cfg; ++n){
        const cfg_t &cfg = m_cfgbase.cfgs[n];
        const string xsd = m_xsd_dir + "\\" + strip_ext(strip_dir(cfg.cn_name)) + ".xsd";
        xsdutil::gen_xsd(cfg, xsd);
    }

    return true;
}

string& xsd_generator::tokit_type_2_xsd_type(enum_tokit_field_type fieldtype)
{
    static string static_xsd_types[fieldtype_max] = {
        "",
        "xs:string",
        "xs:unsignedByte",
        "xs:byte",
        "xs:short",
        "xs:int",
        "xs:long",
        "xs:unsignedByte",
        "xs:unsignedShort",
        "xs:unsignedInt",
        "xs:unsignedLong",
        "xs:float",
        "xs:double"
    };

    return static_xsd_types[fieldtype];
}

File: tokit_go_src/tokit/generator/xsd/xsd_generator.go
///<------------------------------------------------------------------------------
//< @文件名：echo_tool.cpp
//< @作　者：洪坤安
//< @日　期: 2014年11月12日 19:08:42
//< @摘　要:
//< Copyright (c) 2014 Tokit. All rights reserved.
///<------------------------------------------------------------------------------

// xsd_generator
package xsd

import (
	"tokit/generator"
	"tokit/tool"
)

var g_xsd_types = [generator.Fieldtype_max + 1]string{"", "xs:string", "xs:unsignedByte", "xs:byte", "xs:short", "xs:int", "xs:long", "xs:unsignedByte", "xs:unsignedShort", "xs:unsignedInt", "xs:unsignedLong", "xs:float", "xs:double"}

type Xsd_generator struct {
	M_cfgbase *generator.Cfgbase_t
	M_xsd_dir string
}

func (x *Xsd_generator) Generate() {
	if false == tool.Exist(&x.M_xsd_dir) {
		tool.EchoErr("错误: <%s>路径不存在", x.M_xsd_dir)
		return
	}

	x.gen_xsds()
}

func (x *Xsd_generator) gen_xsds() {

	for _, v := range x.M_cfgbase.Cfgs {
		var xsd string = x.M_xsd_dir + "\\" + v.Cn_name + ".xsd"
		gen_xsd(&v, &xsd)
	}
}

func gen_elem(cfg *generator.Cfg_t) string {
	var s string
	s += "<xs:element name=\"" + cfg.En_name + "s\">\n"
	s += "  <xs:complexType>\n"
	s += "    <xs:sequence minOccurs=\"0\" maxOccurs=\"unbounded\">\n"
	s += "      <xs:element name=\"" + cfg.En_name + "\">\n"
	s += "        <xs:complexType>\n"
	s += "			<xs:sequence>\n"

	for _, v := range cfg.Fields {
		s += "            <xs:attribute name=\"" + v.En_name + "\" type=\"" + g_xsd_types[v.Fieldtype] + "\"/>\n"
	}

	s += "          </xs:sequence>\n"
	s += "        </xs:complexType>\n"
	s += "      </xs:element>\n"
	s += "    </xs:sequence>\n"
	s += "  </xs:complexType>\n"
	s += "</xs:element>\n"

	return s
}

func gen_body(cfg *generator.Cfg_t) string {
	var s string
	s += "<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n"
	s += "<xs:schema xmlns:xs=\"http://www.w3.org/2001/XMLSchema\">\n"
	s += gen_elem(cfg)
	s += "</xs:schema>\n"
	return s
}

func gen_xsd(cfg *generator.Cfg_t, xsd *string) bool {
	var text string = gen_body(cfg)
	tool.WriteFile(xsd, &text)
	tool.EchoOk("导出[%s] -> [%s]成功", cfg.Cn_name, *xsd)

	return true
}
